util/svdgen.py

```diff
@@ -0,0 +1,604 @@
+            bitRange            = bitfield(bits['bitinfo']),
+            access              = access,
+            readAction          = readAction,
+            modifiedWriteValues = modifiedWriteValues)
+
+def register(reg: hjson, base=0) -> ET.Element:
+    """Convert a standard register definition into a <register> node.
+    Most of the work was previously done during validation and loading
+    the HJSON file; what's left is mostly transliteration."""
+
+    # To define registers with smaller widths the HJSON includes a
+    # single field in the register; that field has a single element
+    # "bits". During `reggen.validate` the field is expanded with data
+    # gleaned from the parent register.
+    #
+    # SVD has different semantics. Instead the register may have a
+    # `<size>` element set specifying the width. (If not set, the width
+    # is inherited from the top-level <device>.)
+    #
+    # When generating a register we must detect and flatten the field
+    # into the register. This occurs when there is a single field whose
+    # name matches the register and whose bit offset is zero.
+    fields = reg.get('fields')
+    flatten = fields != None and \
+            len(fields) == 1 and \
+            fields[0]['name'] == reg['name'] and \
+            fields[0]['bitinfo'][2] == 0
+
+    size = flatten and fields[0]['bitinfo'][1] or None
+
+    # Ignore an empty or flattened <fields>
+    def regfields(flatten: bool, fields: [hjson]) -> ET.Element:
+        if not flatten and fields != None:
+            yield create('fields', *map(field, fields))
+
+    (access, readAction, modifiedWriteValues) = swaccess(reg)
+    return create('register',
+            name                = reg['name'],
+            description         = reg['desc'],
+            addressOffset       = hex(reg['genoffset'] - base),
+            size                = size,
+            mask                = value(reg.get('genbitsused')),
+            resetValue          = value(reg.get('genresval')),
+            resetMask           = value(reg.get('genresmask')),
+            access              = access,
+            readAction          = readAction,
+            modifiedWriteValues = modifiedWriteValues,
+            *regfields(flatten, fields))
+
+def window(window: hjson) -> ET.Element:
+    """This should generate an SVD <register> element, likely with a set
+    of `dimElementGroup` subelements. Currently, it just returns a comment
+    documenting the omission."""
+
+    window = create(register(window),
+            dim          = window['items'],
+            dimIncrement = value(int(window['genvalidbits']/8)))
+
+    # SVD files require the magic string "%s" in the register to generate
+    # a numbered offset into the window. HJSON doesn't have this; append
+    # it as necessary.
+    name = window.find('./name')
+    if name.text.find('%s') < 0:
+        name.text += '%s'
+
+    return window
+
+def multireg(multi: [hjson]) -> [ET.Element]:
```

Comment:
You're correct, that's what it does now. This function changed over time; at one point it generated a series of <register> elements rather than a parent <cluster> with <register> children. It can be dramatically simplified now.